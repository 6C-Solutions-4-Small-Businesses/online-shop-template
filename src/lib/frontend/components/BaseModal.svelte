<script lang="ts">
    import Button from '$lib/frontend/components/Button.svelte'
    import type {SvelteComponent} from 'svelte'
    import CloseCircleIcon from '~icons/mdi/close-circle'

    export let title: string | undefined
    export let body: string | undefined
    export let isHtml: boolean = false
    export let isFull = false
    export let background = 'bg-white'
    export let parent: SvelteComponent
    export let buttonTextCancel: string | undefined = undefined
    export let buttonTextSubmit: string | undefined = undefined
    export let submitButtonDisabled = false
    export let onSubmitClickHandler: () => Promise<void>
    export let onCancelClickHandler: (() => void) | undefined = undefined
</script>

<div class={`${background} ${isFull ? 'w-screen h-screen overflow-scroll pb-5' : 'w-full'} flex flex-col gap-7 items-center rounded-md`}>
    <header
            class=" w-full flex  justify-between items-center text-md font-bold uppercase bg-green-50 h-14 text-center p-3 text-black rounded-t-md">
        <p>{title }</p>
        <button class="w-8 h-8 flex items-center justify-center rounded-full" data-testid="close-modal-button"
                on:click={onCancelClickHandler ?? parent.onClose}>
            <CloseCircleIcon/>
        </button>
    </header>
    {#if isHtml}
        <div class="px-3 text-slate-600">{@html body}</div>
    {:else if body}
        <p class="w-11/12 text-start text-slate-600">{body}</p>
    {/if}
    <slot/>
    {#if buttonTextCancel || buttonTextSubmit}
        <footer class="w-11/12 flex flex-row md:justify-between gap-2 mb-5">
            {#if buttonTextCancel}
                <Button
                        background="bg-white hover:bg-orange-50"
                        classNames="w-full md:w-2/3 btn {parent.buttonNeutral}  h-12"
                        id="cancel-button"
                        onClick={onCancelClickHandler ?? parent.onClose}>
                    {buttonTextCancel}
                </Button>
            {/if}
            {#if buttonTextSubmit}
                <Button
                        background="bg-orange-50 hover:bg-white"
                        border="border border-orange-500"
                        classNames="w-full md:w-2/3 shadow-sm h-12 btn {parent.buttonPositive}"
                        disabled={submitButtonDisabled}
                        id="continue-button"
                        onClick={onSubmitClickHandler}
                        text="text-orange-500"
                >
                    {buttonTextSubmit}
                </Button>
            {/if}
        </footer>
    {/if}
</div>
