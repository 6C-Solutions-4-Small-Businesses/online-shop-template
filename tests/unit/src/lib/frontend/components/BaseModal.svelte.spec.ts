import {afterEach, beforeEach, describe, expect} from 'vitest'
import '@testing-library/jest-dom'
import {render, type RenderResult, screen as vScreen, waitFor} from '@testing-library/svelte'
import BaseModal from '$lib/frontend/components/BaseModal.svelte'
import type {SvelteComponent} from 'svelte'
import {mock} from 'vitest-mock-extended'

describe('BaseModal component', () => {
    const title = 'Title'
    const body = 'Body'
    const buttonTextCancel = 'annuler'
    const buttonTextSubmit = 'soumettre'
    const parentOnModalClosed = vi.fn()
    const onSubmitClickHandler = vi.fn()
    const onCancelClickHandler = vi.fn()
    let view: RenderResult<BaseModal, typeof import('@testing-library/dom/types/queries')>

    beforeEach(() => {
        view = render(BaseModal, {
            props: {
                title,
                body,
                buttonTextCancel,
                buttonTextSubmit,
                background: 'bg-gray-500 bg-opacity-50',
                parent: mock<SvelteComponent>(),
                onSubmitClickHandler,
                onCancelClickHandler,
            },
        })
    })

    describe('Structure', () => {
        it('should have match snapshot', () => {
            expect(view.container).toMatchSnapshot()
        })

        it('should contains a title', () => {
            const titleElement = vScreen.getByText(title)
            expect(titleElement).toBeInTheDocument()
        })

        it('should contains a body', () => {
            const bodyElement = vScreen.getByText(body)
            expect(bodyElement).toBeInTheDocument()
        })

        it('should contains a cancel button', () => {
            const cancelButton = vScreen.getByText(buttonTextCancel)
            expect(cancelButton).toBeInTheDocument()
        })

        it('should contains a submit button', () => {
            const submitButton = vScreen.getByText(buttonTextSubmit)
            expect(submitButton).toBeInTheDocument()
        })
    })

    describe('Behavior', () => {
        it('should call "onSubmitClickHandler" when submit button is clicked', async () => {
            const submitButton = vScreen.getByText(buttonTextSubmit)
            submitButton.click()

            await waitFor(() => expect(onSubmitClickHandler).toHaveBeenCalled())
        })

        it('should call "parentOnModalClosed" when cancel button is clicked if "onCancelClickHandler" is not set', async () => {
            rerender()

            const cancelButton = vScreen.getByText(buttonTextCancel)
            cancelButton.click()

            await waitFor(() => expect(parentOnModalClosed).toHaveBeenCalled())
        })

        it('should call "onCancelClickHandler" if defined when cancel button is clicked', () => {
            const cancelButton = vScreen.getByText(buttonTextCancel)
            cancelButton.click()

            expect(onCancelClickHandler).toHaveBeenCalled()
        })

        it('should display body has html if "isHtml" is true', () => {
            rerender({ isHtml: true })

            const bodyElement = vScreen.getByText('Body')
            expect(bodyElement).toBeInTheDocument()
        })
    })

    function rerender(properties: { isHtml: boolean } = { isHtml: false }) {
        view.unmount()
        view = render(BaseModal, {
            props: {
                title,
                body: '<div>Body</div>',
                buttonTextCancel,
                buttonTextSubmit,
                parent: mock<SvelteComponent>({
                    onClose: parentOnModalClosed,
                }),
                onSubmitClickHandler,
                isHtml: properties.isHtml,
            },
        })
    }

    afterEach(() => {
        view.unmount()
        vi.clearAllMocks()
    })
})
