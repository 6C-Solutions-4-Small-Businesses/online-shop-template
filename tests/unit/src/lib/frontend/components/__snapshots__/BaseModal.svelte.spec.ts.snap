// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseModal component > Structure > should have match snapshot 1`] = `
<body>
  <div>
    <div
      class="bg-gray-500 bg-opacity-50 w-full flex flex-col gap-7 items-center rounded-md"
    >
      <header
        class="w-full flex justify-between items-center text-md font-bold uppercase bg-green-50 h-14 text-center p-3 text-black rounded-t-md"
      >
        <p>
          Title
        </p>
         
        <button
          class="w-8 h-8 flex items-center justify-center rounded-full"
          data-testid="close-modal-button"
        >
          <svg
            height="1.2em"
            viewBox="0 0 24 24"
            width="1.2em"
          >
            <path
              d="M12 2c5.53 0 10 4.47 10 10s-4.47 10-10 10S2 17.53 2 12S6.47 2 12 2m3.59 5L12 10.59L8.41 7L7 8.41L10.59 12L7 15.59L8.41 17L12 13.41L15.59 17L17 15.59L13.41 12L17 8.41z"
              fill="currentColor"
            />
          </svg>
          <!--&lt;Close-circle&gt;-->
        </button>
      </header>
       
      <p
        class="w-11/12 text-start text-slate-600"
      >
        Body
      </p>
       
       
      <footer
        class="w-11/12 flex flex-row md:justify-between gap-2 mb-5"
      >
        <button
          class="flex justify-around items-center uppercase rounded-sm font-bold w-full md:w-2/3 btn function(...s) {
    let r = A(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let m = r.next.shift();
    if (m) {
      r.results.push(m);
      let [l, o] = m;
      if (l === "ok")
        return o;
      throw o;
    }
    let p, d = "ok";
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), d = "ok";
      } catch (l) {
        throw p = l, d = "error", r.results.push([d, l]), l;
      }
    let a = [d, p];
    if (b(p)) {
      let l = p.then((o) => a[1] = o).catch((o) => {
        throw a[0] = "error", a[1] = o, o;
      });
      Object.assign(l, p), p = l;
    }
    return r.results.push(a), p;
  }  h-12 bg-white hover:bg-orange-50 border text-black"
          data-testid="cancel-button"
          id="cancel-button"
        >
          annuler
        </button>
        <!--&lt;Button&gt;-->
         
        <button
          class="flex justify-around items-center uppercase rounded-sm font-bold w-full md:w-2/3 shadow-sm h-12 btn function(...s) {
    let r = A(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let m = r.next.shift();
    if (m) {
      r.results.push(m);
      let [l, o] = m;
      if (l === "ok")
        return o;
      throw o;
    }
    let p, d = "ok";
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), d = "ok";
      } catch (l) {
        throw p = l, d = "error", r.results.push([d, l]), l;
      }
    let a = [d, p];
    if (b(p)) {
      let l = p.then((o) => a[1] = o).catch((o) => {
        throw a[0] = "error", a[1] = o, o;
      });
      Object.assign(l, p), p = l;
    }
    return r.results.push(a), p;
  } bg-orange-50 hover:bg-white border border-orange-500 text-orange-500"
          data-testid="continue-button"
          id="continue-button"
        >
          soumettre
        </button>
        <!--&lt;Button&gt;-->
      </footer>
    </div>
    <!--&lt;BaseModal&gt;-->
  </div>
</body>
`;
